testutil/gulp-mocha-runner.js

```diff
@@ -44,42 +46,73 @@ var plugin = function (options) {
-        phantom.stdout.setEncoding('utf8');
-        phantom.stdout.on('data', function(data) {
-            var msgStr = data.toString().trim();
-            
-            if (msgStr.charAt(0) === "[") {
-                var msgObj;
+        (async function() {
+            async function launchChrome() {
+                return await chromeLauncher.launch({
+                    chromeFlags: [
+                        "--disable-background-timer-throttling",
+                        "--disable-default-apps",
+                        "--disable-device-discovery-notifications",
+                        "--disable-gpu",
+                        "--disable-popup-blocking",
+                        "--disable-renderer-backgrounding",
+                        "--disable-translate",
+                        "--headless",
+                        "--no-default-browser-check",
+                        "--no-first-run",
+                        "--autoplay-policy=no-user-gesture-required",
+                        "--allow-file-access-from-files"
+                    ],
+                    logLevel: "error"
+                });
+            }
+
+            const chrome = await launchChrome();
+            const protocol = await CDP({port: chrome.port});
+            const {
+                DOM,
+                Page,
+                Runtime,
+                Console
+            } = protocol;
+            await Promise.all([
+                DOM.enable(),
+                Page.enable(),
+                Runtime.enable(),
+                Console.enable()
+            ]);
+
+            Runtime.exceptionThrown((exception) => {
+                let message = exception.exceptionDetails.exception.description;
+                // Don't log infinite loop exceptions,
+                // those indicate the system is working correctly
+                if (message.indexOf("KA_INFINITE_LOOP") !== -1 ||
+                    message.indexOf("testingRuntimeErrors") !== -1) {
+                    return;
+                }
+                errors.push(message);
+            });
+
+            Runtime.consoleAPICalled(({logType, args}) => {
+
+                var logMessage = args[0].value;
+                var logMessageObj;
                 try {
-                    msgObj = JSON.parse(msgStr);
+                    logMessageObj = JSON.parse(logMessage);
                 } catch(e) {
-                    // sometimes we get malformed JSON, just ignore it
-                    if (msgStr.indexOf("pass") > 0) {
-                        process.stdout.write("\x1b[32m.\x1b[0m"); // green
-                    } else if (msgStr.indexOf("fail") > 0) {
-                        process.stdout.write("\x1b[31mF\x1b[0m"); // red
-                        failures.push({
-                            fullTitle: "unknown",
-                            message: "unknown"
-                        });
-                    }
+                    messages.push(unmirror(args[0]));
```

Comment:
So `mirror` is a serialization format that Chrome uses for JS objects.  It looks like sometimes we JSON strings and other times we get these mirror objects.  Could you add a comment about when we might expect each of these possible formats?

Reply:
Good point, this part was confusing for me as well. Adding a comment.